File: Euler_inverso.m
%% 10 años de vueltas
n = 1000000;
t = 2*pi*365*10; 
dt = t/n; 
B = eye(2)-dt*[0 1;-1 0]; 
U = zeros(2,n+1);
U(:,1) = [1 0]';
for i = 1:n
    U(:,i+1) = B\U(:,i);
end
plot(U(1,:),U(2,:))
grid on
axis('equal');
xlabel('$x$', 'FontSize',14,'Interpreter','latex')
ylabel('$dx/dt$', 'FontSize',14,'Interpreter','latex')
radio = sqrt(U(1,n+1)^2+U(2,n+1)^2);


%% F en funcion
n = 1000000;
t = 2*pi*365*10; 
dt = t/n; 
U = zeros(2,n+1);
U(:,1) = [1 0]';
for i = 1:n
    U(:,i+1) = U(:,i)+dt*Oscillator(U(:,i));
end

plot(U(1,:),U(2,:))
grid on
axis('equal');
xlabel('$x$', 'FontSize',14,'Interpreter','latex')
ylabel('$dx/dt$', 'FontSize',14,'Interpreter','latex')

function F = Oscillator(U)
F = [U(2); -U(1)];
end 

File: Python/Euler_inverso.py
import math
from XMP_MUSE import euler_implicito
import numpy as np
import matplotlib.pyplot as plt


n = int(1e6)
t = 2*math.pi*365*10
dt = t/n
A = np.array([[0, 1], [-1, 0]])
U = np.empty((2,n+1))
U [:,0] = [1,0]
U = euler_implicito(U,dt,n,A)

plt.figure(figsize = (8,8))
plt.plot(U[0,:], U[1,:])
plt.xlabel(r"$x\ (m)$", fontsize = 14, color = 'black')
plt.ylabel(r"$dx/dt\ (m/s)$", fontsize = 14, color = 'black')
plt.grid(True)
plt.grid(color = '0.5', linestyle = '--', linewidth = '1')
plt.axis('tight')
plt.show()

radio = math.sqrt(U[0,n]**2+U[1,n]**2)
print('Radio de la órbita =', radio, 'm')